components/datetime/src/arithmetic.rs

```diff
@@ -92,3 +143,401 @@ pub fn is_top_of_hour(pattern: &Pattern, minute: u8, second: u8) -> bool {
         Some(TimeGranularity::Seconds) => minute + second == 0,
     }
 }
+
+pub mod week_of {
```

Comment:
I feel like this math should be moved into the icu_calendars crate, perhaps as a  utilities module for now, since there already is a bunch of other math there. Not sure if that should be done in this PR, but I think it'll help simplify that